pkg/validate/security_context_linux.go

```diff
@@ -639,6 +588,153 @@ var _ = framework.KubeDescribe("Security Context", func() {
+							SupplementalGroups: []int64{supplementalGroup},
+						},
+					},
+					LogPath: logPath,
+				}
+				containerID := framework.CreateContainer(rc, ic, containerConfig, podID, podConfig)
+
+				By("start container")
+				startContainer(rc, containerID)
+
+				Eventually(func(g Gomega) {
+					containerStatus := getContainerStatus(rc, containerID)
+					g.Expect(containerStatus.State).To(Equal(runtimeapi.ContainerState_CONTAINER_RUNNING))
+					// In testImagePreDefinedGroup,
+					// - its default user is default-user(uid=1000)
+					// - default-user belongs to group-defined-in-image(gid=50000) in /etc/group
+					// And, SupplementalGroupsPolicy is Merge(default)
+					//
+					// Thus, firstly attached process identity of the first container processes should be
+					// - uid: 1000 (RunAsUser)
+					// - gid: 1000 (default group for uid=1000)
+					// - supplementary groups
+					//   - 1000: self
+					//   - 1234: SupplementalGroups
+					//   - 50000: groups defined in the container image (/etc/group)
+					if containerStatus.User != nil && containerStatus.User.Linux != nil {
+						slices.Sort(containerStatus.User.Linux.SupplementalGroups)
+					}
+					g.Expect(containerStatus.User).To(BeEquivalentTo(&runtimeapi.ContainerUser{
+						Linux: &runtimeapi.LinuxContainerUser{
+							Uid:                imagePredefinedGroupUID,
+							Gid:                imagePredefinedGroupUID,
+							SupplementalGroups: []int64{imagePredefinedGroupUID, supplementalGroup, imagePredefinedGroupGID},
+						},
+					}))
+					g.Expect(parseLogLine(podConfig, logPath)).NotTo(BeEmpty())
+				}, time.Minute, time.Second*4).Should(Succeed())
+
+				// $ id -G
+				// 1000 1234 50000
+				expectedOutput := fmt.Sprintf("%d %d %d\n", imagePredefinedGroupUID, supplementalGroup, imagePredefinedGroupGID)
+
+				By("verify groups for the first process of the container")
+				verifyLogContents(podConfig, logPath, expectedOutput, stdoutType)
+
+				By("verify groups for 'exec'-ed process of container")
+				command := []string{"id", "-G"}
+				o := execSyncContainer(rc, containerID, command)
+				Expect(o).To(BeEquivalentTo(expectedOutput))
+			})
+		})
+		When("SupplementalGroupsPolicy=Strict", func() {
+			It("even if the container's primary UID belongs to some groups in the image, runtime should not add SupplementalGroups to them", func() {
+				By("create pod")
+				podID, podConfig, podLogDir = createPodSandboxWithLogDirectory(rc)
+
+				By("create container for security context SupplementalGroups")
+				supplementalGroup := int64(1234)
+				containerName := "container-with-SupplementalGroupsPolicyMerge-" + framework.NewUUID()
+				logPath := containerName + ".log"
+				containerConfig := &runtimeapi.ContainerConfig{
+					Metadata: framework.BuildContainerMetadata(containerName, framework.DefaultAttempt),
+					Image:    &runtimeapi.ImageSpec{Image: testImagePreDefinedGroup},
+					Command:  []string{"sh", "-c", "id -G; while :; do sleep 1; done"},
```

Comment:
Would this work in place of the loop?

```suggestion
					Command:  []string{"sh", "-c", "id -G; sleep infinity"},
```

Reply:
I realised this is a code that existed prior... hence I apologise for asking about it. However, perhaps we can make things bit better. :smile:

Reply:
never mind. fixed in c9e3de68